realsense2_camera/src/realsense_node_factory.cpp

```diff
@@ -244,47 +266,89 @@ void RealSenseNodeFactory::onInit()
     	privateNh.param("usb_port_id", _usb_port_id, std::string(""));
     	privateNh.param("device_type", _device_type, std::string(""));
 
+		std::string timeout_action;
+		privateNh.param("timeout_action", timeout_action, std::string("warn"));
+		if (timeout_action == "warn")
+		{
+			_timeout_action = TIMEOUT_WARN;
+		}
+		else if (timeout_action == "reset")
+		{
+			_timeout_action = TIMEOUT_RESET;
+		}
+		else if (timeout_action == "shutdown")
+		{
+			_timeout_action = TIMEOUT_SHUTDOWN;
+		}
+		else
+		{
+			_timeout_action = TIMEOUT_WARN;
+			ROS_WARN("Invalid timeout_action (warn|reset|shutdown): %s", timeout_action.c_str());
+		}
+
 		std::string rosbag_filename("");
 		privateNh.param("rosbag_filename", rosbag_filename, std::string(""));
 		if (!rosbag_filename.empty())
 		{
-			{
-				ROS_INFO_STREAM("publish topics from rosbag file: " << rosbag_filename.c_str());
-				auto pipe = std::make_shared<rs2::pipeline>();
-				rs2::config cfg;
-				cfg.enable_device_from_file(rosbag_filename.c_str(), false);
-				cfg.enable_all_streams();
-				pipe->start(cfg); //File will be opened in read mode at this point
-				_device = pipe->get_active_profile().get_device();
-				_serial_no = _device.get_info(RS2_CAMERA_INFO_SERIAL_NUMBER);
-			}
-			if (_device)
-			{
-				StartDevice();
-			}
+			ROS_INFO_STREAM("publish topics from rosbag file: " << rosbag_filename.c_str());
+			auto pipe = std::make_shared<rs2::pipeline>();
+			rs2::config cfg;
+			cfg.enable_device_from_file(rosbag_filename.c_str(), false);
+			cfg.enable_all_streams();
+			pipe->start(cfg); //File will be opened in read mode at this point
+			_device = pipe->get_active_profile().get_device();
+			_realSenseNode = std::shared_ptr<BaseRealSenseNode>(new BaseRealSenseNode(nh, privateNh, _device, _serial_no));
+			_realSenseNode->publishTopics();
+			_realSenseNode->registerDynamicReconfigCb(nh);
 		}
 		else
 		{
-			privateNh.param("initial_reset", _initial_reset, false);
-
+			_is_alive = true;
 			_query_thread = std::thread([=]()
 						{
+							ROS_DEBUG("Waiting for device...");
+
+ 
 							std::chrono::milliseconds timespan(6000);
 							while (_is_alive && !_device)
 							{
-								getDevice(_ctx.query_devices());
+								ROS_DEBUG("Checking for device...");
+								_device = getDevice();
+ 								// getDevice(_ctx.query_devices());  // This line is obtained from the latest upstream.
 								if (_device)
 								{
-									std::function<void(rs2::event_information&)> change_device_callback_function = [this](rs2::event_information& info){change_device_callback(info);};
-									_ctx.set_devices_changed_callback(change_device_callback_function);
-									StartDevice();
+									ROS_DEBUG("got device");
+									if (_initial_reset)
```

Comment:
New new getDevice() method appears to handle the initial reset logic, so I think this section can be removed